backend/test/v2/integration/upgrade_test.go

```diff
@@ -0,0 +1,601 @@
+	require.Nil(t, err)
+	// During upgrade, default pipelines may be installed, so we only verify the
+	// 4 oldest pipelines here.
+	assert.True(t, len(pipelines) >= 4)
+	assert.Equal(t, "arguments-parameters.yaml", pipelines[0].DisplayName)
+	assert.Equal(t, "sequential", pipelines[1].DisplayName)
+	assert.Equal(t, "zip-arguments-parameters", pipelines[2].DisplayName)
+	assert.Equal(t, "arguments.pipeline.zip", pipelines[3].DisplayName)
+
+	/* ---------- Verify pipeline spec ---------- */
+	pipelineVersions, totalSize, _, err := s.pipelineClient.ListPipelineVersions(
+		&params.ListPipelineVersionsParams{
+			PipelineID: pipelines[0].PipelineID,
+		})
+	require.Nil(t, err)
+	assert.Equal(t, totalSize, 1)
+	pipelineVersion, err := s.pipelineClient.GetPipelineVersion(&params.GetPipelineVersionParams{PipelineID: pipelines[0].PipelineID, PipelineVersionID: pipelineVersions[0].PipelineVersionID})
+	require.Nil(t, err)
+	bytes, err := ioutil.ReadFile("../resources/arguments-parameters.yaml")
+	expected_bytes, err := yaml.YAMLToJSON(bytes)
+	require.Nil(t, err)
+	actual_bytes, err := json.Marshal(pipelineVersion.PipelineSpec)
+	require.Nil(t, err)
+	// Override pipeline name, then compare
+	assert.Equal(t, string(expected_bytes), strings.Replace(string(actual_bytes), "pipeline/arguments-parameters.yaml", "whalesay", 1))
+}
+
+func (s *UpgradeTests) PrepareRuns() {
+	t := s.T()
+
+	helloWorldPipeline := s.getHelloWorldPipeline(true)
+	helloWorldExperiment := s.getHelloWorldExperiment(true)
+	if helloWorldExperiment == nil {
+		helloWorldExperiment = s.createHelloWorldExperiment()
+	}
+
+	hello2 := s.getHelloWorldExperiment(true)
+	require.Equal(t, hello2, helloWorldExperiment)
+
+	/* ---------- Create a new hello world run by specifying pipeline ID ---------- */
+	createRunRequest := &runParams.CreateRunParams{Body: &run_model.V2beta1Run{
+		DisplayName:  "hello world",
+		Description:  "this is hello world",
+		ExperimentID: helloWorldExperiment.ExperimentID,
+		PipelineVersionReference: &run_model.V2beta1PipelineVersionReference{
+			PipelineID:        helloWorldPipeline.PipelineID,
+			PipelineVersionID: helloWorldPipeline.PipelineVersionID,
+		},
+	}}
+	_, err := s.runClient.Create(createRunRequest)
+	require.Nil(t, err)
+}
+
+func (s *UpgradeTests) VerifyRuns() {
+	t := s.T()
+
+	/* ---------- List the runs, sorted by creation time ---------- */
+	runs, _, _, err := test.ListRuns(
+		s.runClient,
+		&runParams.ListRunsParams{SortBy: util.StringPointer("created_at")},
+		s.resourceNamespace)
+	require.Nil(t, err)
+	assert.True(t, len(runs) >= 1)
+	assert.Equal(t, "hello world", runs[0].DisplayName)
+
+	/* ---------- Get hello world run ---------- */
+	helloWorldRunDetail, err := s.runClient.Get(&runParams.GetRunParams{RunID: runs[0].RunID})
+	require.Nil(t, err)
+	assert.Equal(t, "hello world", helloWorldRunDetail.DisplayName)
+	assert.Equal(t, "this is hello world", helloWorldRunDetail.Description)
```

Comment:
I think we verified more fields of the returned run in v1 ([ref](https://github.com/kubeflow/pipelines/blob/ccbac97a7bc3f3d65151fd46d7714162e15cbfc5/backend/test/integration/upgrade_test.go#L568)). We can mask the fields that vary in each test with definite values, and then make the comparison.

Reply:
The v1 code actually doesn't verify more, if you look closely, the pattern is basically:

```golang
func check(t, actual) {
   expected = Run{
       Foo: actual.Foo,
       Bar: actual.Bar,
   }
   assert.Equal(t, expected, actual)
}
```

In a word, they are constructing the `expected` using the `actual`, then compare the two. There's actually little verification value provided with this pattern.

Reply:
I think you are right, much of the fields are overwritten by the returned values. But the v1 code also verifies that some fields are not empty. Maybe we can also check if fields like pipeline spec, exp id etc. are non-empty?

Reply:
pipeline spec would be empty in this case because the run was created using pipeline version id, which is a behavior change from v1. We do have API integration test that verifies when pipeline spec shouldn't be empty (for run created using pipeline spec).
For experiment id, I can certainly verify that it's not empty, but I think it's no different from verifying any other field being non-empty and we don't do this extensive non-empty check for other fields. In v1, the specialty with experiment id is it belongs to resource reference, so it makes sense to verify that is constructed correctly. In v2, we no longer have resource references.

Reply:
Thank you for explaining! This makes a lot of sense.